### PR TITLE
ObjFifo unroll dependency fixes

### PR DESCRIPTION
**Pull request by AndraBisca**
This fix introduces a few fixes in the ObjectFifo lowering's for-loop unrolling algorithm:
- make the for-loop operations dependency detection recursive to account for nested loops,
- add operations that were duplicated as a result of duplicating a nested for-loop to the list of duplicate operations.

As well as a fix for replacing ObjectFifo acquire / release operations:
- account for acquire / release operations that are inside two different block regions.

### CODE REVIEW

File: include/aie/Dialect/AIE/IR/AIE.td [hunk 1]
```diff
@@ -1370,26 +1370,27 @@ def AIE_ShimDMAAllocationOp : AIE_Op<"shimDMAAllocation", [HasParent<"DeviceOp">
...
-    of its object pool.
-
-    For the producer and for each consumer, a different size (i.e., element number) can be specified. This will 
-    take effect in the case of consumers placed on tiles non-adjacent to the producer. Otherwise, the producer 
-    size will be applied. If a single size is specified, it will be applied to both producer and consumers.
-
-    This operation is then converted by the AIEObjectFifoStatefulTransformPass into buffers and their associated 
-    locks. The pass also establishes Flow and DMA operations between the producer and consumer tiles if they are
+    The `aie.objectFifo` operation creates a circular buffer established between a producer and one or 
+    more consumers, which are `aie.tile` operations. The aie.objectFifo instantiates the given number of
```
Comment (keryell) [inline include/aie/Dialect/AIE/IR/AIE.td:1374]:
```suggestion
    more consumers, which are `aie.tile` operations. The `aie.objectFifo` instantiates the given number of 
```

File: include/aie/Dialect/AIE/IR/AIE.td [hunk 2]
```diff
@@ -1370,26 +1370,27 @@ def AIE_ShimDMAAllocationOp : AIE_Op<"shimDMAAllocation", [HasParent<"DeviceOp">
...
-    For the producer and for each consumer, a different size (i.e., element number) can be specified. This will 
-    take effect in the case of consumers placed on tiles non-adjacent to the producer. Otherwise, the producer 
-    size will be applied. If a single size is specified, it will be applied to both producer and consumers.
-
-    This operation is then converted by the AIEObjectFifoStatefulTransformPass into buffers and their associated 
-    locks. The pass also establishes Flow and DMA operations between the producer and consumer tiles if they are
+    The `aie.objectFifo` operation creates a circular buffer established between a producer and one or 
+    more consumers, which are `aie.tile` operations. The aie.objectFifo instantiates the given number of 
+    buffers (of given output type) and their locks in the Memory Module of the appropriate tile(s) after 
+    lowering, based on tile-adjacency. These elements represent the conceptual depth of the objectFifo or,
```
Comment (keryell) [inline include/aie/Dialect/AIE/IR/AIE.td:1376]:
```suggestion
    lowering, based on tile-adjacency. These elements represent the conceptual depth of the `objectFifo` or, 
```

File: include/aie/Dialect/AIE/IR/AIE.td [hunk 3]
```diff
@@ -1370,26 +1370,27 @@ def AIE_ShimDMAAllocationOp : AIE_Op<"shimDMAAllocation", [HasParent<"DeviceOp">
...
+    buffers (of given output type) and their locks in the Memory Module of the appropriate tile(s) after 
+    lowering, based on tile-adjacency. These elements represent the conceptual depth of the objectFifo or, 
+    more specifically, of its object pool.
+
+    For the producer and for each consumer, a different size (i.e., element number) can be specified as an 
+    array of integer values. This will take effect in the case of consumers placed on tiles non-adjacent to 
+    the producer. Otherwise, the producer size will be applied. If a single size is specified, it will be 
+    applied to both producer and consumers.
+
+    This operation is then converted by the AIEObjectFifoStatefulTransformPass into aie.buffers and their associated
```
Comment (keryell) [inline include/aie/Dialect/AIE/IR/AIE.td:1384]:
```suggestion
    This operation is then converted by the `AIEObjectFifoStatefulTransformPass` into `aie.buffers` and their associated 
```

File: include/aie/Dialect/AIE/IR/AIE.td [hunk 4]
```diff
@@ -1370,26 +1370,27 @@ def AIE_ShimDMAAllocationOp : AIE_Op<"shimDMAAllocation", [HasParent<"DeviceOp">
...
 
     1-to-1 tile example:
     ```
-      AIE.objectFifo @of1 (%tile12, { %tile13 }, 4 : i32) : !AIE.objectFifo<memref<16xi32>> 
+      AIE.objectFifo @of1 (%tile12, { %tile23 }, 4 : i32) : !AIE.objectFifo<memref<16xi32>> 
     ```
-    This operation creates an objectFifo between %tile12 and %tile13 of 4 elements, each a buffer of 16 32-bit integers.
-    Note: If there are no ObjectFifoAcquireOps corresponding to this objectFifo on the cores of %tile12 and %tile13, 
+    This operation creates an objectFifo between %tile12 and %tile23 of 4 elements, each a buffer of 16 32-bit integers.
+    Note: If there are no ObjectFifoAcquireOps corresponding to this objectFifo on the cores of %tile12 and %tile23,
```
Comment (keryell) [inline include/aie/Dialect/AIE/IR/AIE.td:1393]:
```suggestion
    This operation creates an `objectFifo` between `%tile12` and `%tile23` of 4 elements, each a buffer of 16 32-bit integers.
    Note: If there are no `ObjectFifoAcquireOps` corresponding to this `objectFifo` on the cores of `%tile12` and `%tile23`, 
```

**Review by jgmelber — APPROVED** [review]

**Review by keryell — APPROVED** [review]
Thanks!

**Comment by AndraBisca** [comment]
> Thanks!

Thank you for taking the time to check the PR out! :)

**Pull request merged**